stl/inc/span

```diff
@@ -0,0 +1,778 @@
+        if constexpr (_Extent != dynamic_extent) {
+            static_assert(_Extent <= dynamic_extent / sizeof(_Ty),
+                "size of span in bytes exceeds std::numeric_limits<size_t>::max()");
+        }
+#if _CONTAINER_DEBUG_LEVEL > 0
+        else {
+            _STL_VERIFY(size() <= dynamic_extent / sizeof(_Ty),
+                "size of span in bytes exceeds std::numeric_limits<size_t>::max()");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return size() * sizeof(element_type);
+    }
+
+    _NODISCARD constexpr bool empty() const noexcept {
+        return size() == 0;
+    }
+
+    // [span.elem] Element access
+    _NODISCARD constexpr reference operator[](const size_type _Off) const noexcept /* strengthened */ {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_Off < size(), "span index out of range");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return _Mydata[_Off];
+    }
+
+    _NODISCARD constexpr reference front() const noexcept /* strengthened */ {
+        if constexpr (_Extent != dynamic_extent) {
+            static_assert(_Extent > 0, "front of empty span");
+        }
+#if _CONTAINER_DEBUG_LEVEL > 0
+        else {
+            _STL_VERIFY(size() > 0, "front of empty span");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return _Mydata[0];
+    }
+
+    _NODISCARD constexpr reference back() const noexcept /* strengthened */ {
+        if constexpr (_Extent != dynamic_extent) {
+            static_assert(_Extent > 0, "back of empty span");
+        }
+#if _CONTAINER_DEBUG_LEVEL > 0
+        else {
+            _STL_VERIFY(size() > 0, "back of empty span");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return _Mydata[size() - 1];
+    }
+
+    _NODISCARD constexpr pointer data() const noexcept {
+        return _Mydata;
+    }
+
+    // [span.iterators] Iterator support
+    _NODISCARD constexpr iterator begin() noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return {_Mydata, _Mydata, _Mydata + size()};
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return {_Mydata};
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr iterator end() noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return {_Mydata + size(), _Mydata, _Mydata + size()};
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return {_Mydata + size()};
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr const_iterator begin() const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return {_Mydata, _Mydata, _Mydata + size()};
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return {_Mydata};
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr const_iterator end() const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return {_Mydata + size(), _Mydata, _Mydata + size()};

```

Comment:
This calls `size()` twice. In recent years, we've been trying to improve non-optimized debug performance in important functions. `span::end` seems important, so I recommend extracting this `size()` call, storing it in a `const size_type`. Then only one function call will be executed.

Reply:
I went with a temporary